internal/testutils/fakegrpclb/server.go

```diff
@@ -0,0 +1,247 @@
+	sOpts       []grpc.ServerOption // gRPC server options.
+	serviceName string              // Service name being load balanced for.
+	servicePort int                 // Service port being load balanced for.
+	shortStream bool                // End balancer stream after sending server list.
+
+	// Values initialized using ServerParams passed to NewServer.
+	backends []*lbpb.Server // Service backends to balance load across.
+	lis      net.Listener   // Listener for grpc connections to the LoadBalancer service.
+
+	// mu guards access to below fields.
+	mu         sync.Mutex
+	grpcServer *grpc.Server // Underlying grpc server.
+	address    string       // Actual listening address.
+
+	stopped chan struct{} // Closed when Stop() is called.
+}
+
+// NewServer creates a new Server with passed in params. Returns a non-nil error
+// if the params are invalid.
+func NewServer(params ServerParams) (*Server, error) {
+	var servers []*lbpb.Server
+	for _, addr := range params.BackendAddresses {
+		ipStr, portStr, err := net.SplitHostPort(addr)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse list of backend address %q: %v", addr, err)
+		}
+		ip := net.ParseIP(ipStr)
+		if ip == nil {
+			return nil, fmt.Errorf("failed to parse ip: %q", ipStr)
+		}
+		port, err := strconv.Atoi(portStr)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert port %q to int", portStr)
+		}
+		logger.Infof("Adding backend ip: %q, port: %d to server list", ip.String(), port)
+		servers = append(servers, &lbpb.Server{
+			IpAddress: ip,
+			Port:      int32(port),
+		})
+	}
+
+	lis, err := net.Listen("tcp", ":"+strconv.Itoa(params.ListenPort))
+	if err != nil {
+		return nil, fmt.Errorf("failed to listen on port %q: %v", params.ListenPort, err)
+	}
+
+	return &Server{
+		sOpts:       params.ServerOptions,
+		serviceName: params.LoadBalancedServiceName,
+		servicePort: params.LoadBalancedServicePort,
+		shortStream: params.ShortStream,
+		backends:    servers,
+		lis:         lis,
+		address:     lis.Addr().String(),
+		stopped:     make(chan struct{}),
+	}, nil
+}
+
+// Serve starts serving the LoadBalancer service on a gRPC server.
+//
+// It returns early with a non-nil error if it is unable to start serving.
+// Otherwise, it blocks until Stop() is called, at which point it returns the
+// error returned by the underlying grpc.Server's Serve() method.
+func (s *Server) Serve() error {
+	if s.grpcServer != nil {
+		return errors.New("Serve() called multiple times")
+	}
+
+	server := grpc.NewServer(s.sOpts...)
+	s.mu.Lock()
+	s.grpcServer = server
+	s.mu.Unlock()
+
+	logger.Infof("Begin listening on %s", s.lis.Addr().String())
+	lbgrpc.RegisterLoadBalancerServer(server, s)
+	return server.Serve(s.lis) // This call will block.
+}
+
+// Stop stops serving the LoadBalancer service and unblocks the preceding call
+// to Serve().
+func (s *Server) Stop() {
+	defer close(s.stopped)
+	s.mu.Lock()
+	if s.grpcServer != nil {
+		s.grpcServer.Stop()
```

Comment:
Set it to nil? It probably doesn't matter unless we want to call Serve() after Stop()